File: packages/expo-webauthn/android/src/main/java/expo/porto/webauthn/WebAuthNSharedTypes.kt
package expo.porto.webauthn

import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import expo.modules.kotlin.types.Enumerable
import expo.porto.webauthn.Base64Utils.toBase64URLString
import org.json.JSONObject
import org.json.JSONArray

// MARK: - Type Aliases and Constants
typealias Base64URLString = String
typealias CredentialIDString = Base64URLString

// MARK: - Enums
enum class PublicKeyCredentialType(val value: String) : Enumerable {
    PUBLIC_KEY("public-key")
}

enum class PublicKeyCredentialAlgorithm(val value: Int) : Enumerable {
    ES256(-7)
}

enum class AuthenticatorTransport(val value: String) : Enumerable {
    USB("usb"),
    NFC("nfc"),
    BLE("ble"),
    HYBRID("hybrid");
}

enum class ResidentKeyPreference(val value: String) : Enumerable {
    DISCOURAGED("discouraged"),
    PREFERRED("preferred"),
    REQUIRED("required")
}

enum class UserVerificationPreference(val value: String) : Enumerable {
    DISCOURAGED("discouraged"),
    PREFERRED("preferred"),
    REQUIRED("required")
}

enum class AttestationConveyancePreference(val value: String) : Enumerable {
    DIRECT("direct"),
    ENTERPRISE("enterprise"),
    INDIRECT("indirect"),
    NONE("none")
}

// MARK: - Common Types
data class PublicKeyCredentialDescriptor(
    @Field
    val id: CredentialIDString,

    @Field
    val type: PublicKeyCredentialType = PublicKeyCredentialType.PUBLIC_KEY,

    @Field
    val transports: List<AuthenticatorTransport>? = null
) : Record {
    internal fun toJSON(): JSONObject = JSONObject().apply {
        put("type", "public-key")
        put("id", id.toBase64URLString())
        transports?.let {
            put("transports", JSONArray(it.map { transport -> transport.value }))
        }
    }
}

data class RelyingParty(
    @Field
    val id: String,

    @Field
    val name: String
) : Record

data class User(
    @Field
    val id: Base64URLString,

    @Field
    val name: String,

    @Field
    val displayName: String
) : Record

File: packages/expo-webauthn/ios/WebAuthNSharedTypes.swift
import ExpoModulesCore
import AuthenticationServices

// MARK: - Type Aliases
/// A string that is base64URL encoded
public typealias Base64URLString = String

/// A string that represents a credential ID in base64URL format
public typealias CredentialIDString = Base64URLString

// MARK: - Enums
/// Represents the COSE algorithm identifier as defined in AuthenticationServices
/// @see https://developer.apple.com/documentation/authenticationservices/ascosealgorithmidentifier
public enum PublicKeyCredentialAlgorithm: Int, Enumerable {
    /// ECDSA with SHA-256
    case es256 = -7
    var asAlgorithmIdentifier: ASCOSEAlgorithmIdentifier {
        switch self {
        case .es256:
            return .ES256
        }
    }
}

public enum PublicKeyCredentialType: String, Enumerable {
    case publicKey = "public-key"
}

/// Represents the transport method for security key credentials
public enum AuthenticatorTransport: String, Enumerable {
    case usb
    case nfc
    case ble
    case hybrid
    
    var asTransport: ASAuthorizationSecurityKeyPublicKeyCredentialDescriptor.Transport? {
        switch self {
        case .ble:
            return .bluetooth
        case .nfc:
            return .nfc
        case .usb:
            return .usb
        default:
            return nil
        }
    }
}

public enum ResidentKeyPreference: String, Enumerable {
    case discouraged
    case preferred
    case required
    
    var asPreference: ASAuthorizationPublicKeyCredentialResidentKeyPreference {
        switch self {
        case .discouraged:
            return .discouraged
        case .preferred:
            return .preferred
        case .required:
            return .required
        }
    }
}

public enum UserVerificationPreference: String, Enumerable {
    case discouraged
    case preferred
    case required
    
    var asPreference: ASAuthorizationPublicKeyCredentialUserVerificationPreference {
        switch self {
        case .discouraged:
            return .discouraged
        case .preferred:
            return .preferred
        case .required:
            return .required
        }
    }
}

public enum AttestationConveyancePreference: String, Enumerable {
    case direct
    case enterprise
    case indirect
    case none
    
    var asKind: ASAuthorizationPublicKeyCredentialAttestationKind {
        switch self {
        case .direct:
            return .direct
        case .enterprise:
            return .enterprise
        case .indirect:
            return .indirect
        case .none:
            return .none
        }
    }
}

// MARK: - Common Types
public struct PublicKeyCredentialDescriptor: Record {
    
    @Field 
    var type: PublicKeyCredentialType = .publicKey
    
    @Field 
    var id: CredentialIDString
    
    @Field 
    var transports: [AuthenticatorTransport]?
    
    func toPlatformDescriptor() throws -> ASAuthorizationPlatformPublicKeyCredentialDescriptor {
        guard let credentialData = Data(base64URLEncoded: id) else {
            throw InvalidCredentialException()
        }
        return ASAuthorizationPlatformPublicKeyCredentialDescriptor(credentialID: credentialData)
    }
    
    func toSecurityKeyDescriptor() throws -> ASAuthorizationSecurityKeyPublicKeyCredentialDescriptor {
        guard let credentialData = Data(base64URLEncoded: id) else {
            throw InvalidCredentialException()
        }
        
        let descriptor = ASAuthorizationSecurityKeyPublicKeyCredentialDescriptor(
            credentialID: credentialData,
            transports: ASAuthorizationSecurityKeyPublicKeyCredentialDescriptor.Transport.allSupported
        )
        
        if let transports = transports, !transports.isEmpty {
            let validTransports = transports.compactMap { $0.asTransport }
            if !validTransports.isEmpty {
                descriptor.transports = validTransports
            }
        }
        
        return descriptor
    }
}

public struct RelyingParty: Record {
    @Field var id: String
    @Field var name: String
}

public struct User: Record {
    @Field var id: Base64URLString
    @Field var name: String
    @Field var displayName: String
} 
